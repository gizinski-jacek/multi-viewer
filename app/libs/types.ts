export type Hosts =
	| 'youtube'
	| 'youtube-playlist'
	| 'twitch'
	| 'twitch-vod'
	| 'dailymotion'
	| 'dailymotion-playlist'
	| 'vimeo'
	| '';

export interface VideoData {
	host: Hosts;
	id: string;
	title: string | null;
	channelId: string | null;
	channelName: string;
	iFrameSrcId: string;
	livestreamChat: boolean;
	thumbnailUrl: string | null;
}

export interface YoutubeResponse {
	kind: string;
	etag: string;
	nextPageToken: string;
	prevPageToken: string;
	pageInfo: {
		totalResults: number;
		resultsPerPage: number;
	};
	items: YoutubeVideo[];
}

export interface YoutubeVideo {
	kind: string;
	etag: string;
	id: string;
	snippet: {
		publishedAt: string;
		channelId: string;
		title: string;
		description: string;
		thumbnails: {
			[key: string]: {
				url: string;
				width: number;
				height: number;
			};
		};
		channelTitle: string;
		tags: string[];
		categoryId: string;
		liveBroadcastContent: string;
		defaultLanguage: string;
		localized: {
			title: string;
			description: string;
		};
		defaultAudioLanguage: string;
	};
}

export interface TwitchAuth {
	access_token: string;
	expires_in: number;
	token_type: string;
}

export interface TwitchStream {
	data: {
		id: string;
		user_id: string;
		user_login: string;
		user_name: string;
		game_id: string;
		game_name: string;
		type: string;
		title: string;
		viewer_count: number;
		started_at: string;
		language: string;
		thumbnail_url: string;
		tag_ids: string[];
		tags: string[];
		is_mature: boolean;
	}[];
}
export interface TwitchVOD {
	data: {
		id: string;
		stream_id: null;
		user_id: string;
		user_login: string;
		user_name: string;
		title: string;
		description: string;
		created_at: string;
		published_at: string;
		url: string;
		thumbnail_url: string;
		viewable: string;
		view_count: number;
		language: string;
		type: string;
		duration: string;
		muted_segments: {
			duration: number;
			offset: number;
		}[];
	}[];
	pagination: {};
}

export interface DailymotionVideo {
	id: string;
	title: string;
	'owner.id': string;
	'owner.username': string;
}

export interface DailymotionPlaylist {
	id: string;
	name: string;
	'owner.id': string;
	'owner.username': string;
}

export interface VimeoVideo {
	uri: string;
	name: string;
	type: string;
	user: {
		uri: string;
		name: string;
		link: string;
	};
}
