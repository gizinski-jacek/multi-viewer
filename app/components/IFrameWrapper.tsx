export function IFrameWrapper({ src }: { src: string }) {
	return (
		<iframe
			width='100%'
			height='100%'
			src={src}
			title='Twitch video player'
			allow='accelerometer; encrypted-media; gyroscope; picture-in-picture; web-share'
			sandbox='allow-modals allow-scripts allow-same-origin allow-popups allow-popups-to-escape-sandbox allow-storage-access-by-user-activation'
			referrerPolicy='strict-origin-when-cross-origin'
			allowFullScreen
		></iframe>
	);
}
