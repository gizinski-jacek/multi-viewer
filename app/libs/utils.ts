import axios, { AxiosError, AxiosResponse } from 'axios';
import { Hosts, VideoData } from './types';
import { NextResponse } from 'next/server';

export function extractVideoId(host: Hosts, string: string): string {
	if (!host) throw new Error('Select video host');
	const str =
		string[string.length - 1] === '/'
			? string.slice(0, string.length - 1)
			: string;
	switch (host) {
		case 'youtube':
			let idYT = str
				.replace('https://youtu.be/', '')
				.replace('https://www.youtube.com/watch?v=', '')
				.replace('https://www.youtube.com/live/', '');
			if (idYT.indexOf('&list=') > 1) {
				idYT = idYT.slice(0, idYT.indexOf('&list='));
			}
			if (idYT.indexOf('?si=') > 1) {
				idYT = idYT.slice(0, idYT.indexOf('?si='));
			}
			return idYT;
		case 'youtube-playlist':
			let idYTP = str
				.replace('https://youtube.com/playlist?list=', '')
				.replace('https://www.youtube.com/playlist?list=', '');
			if (idYTP.includes('?si=')) {
				idYTP = idYTP.slice(0, idYTP.indexOf('?si='));
			}
			return idYTP;
		case 'twitch':
		case 'twitch-vod':
		case 'dailymotion':
		case 'dailymotion-playlist':
		case 'vimeo':
			return str.slice(str.lastIndexOf('/') + 1);
		default:
			throw new Error('Unsupported host or incorrect ID');
	}
}

export async function getVideoData(
	host: Hosts,
	id: string
): Promise<VideoData> {
	try {
		if (!host) {
			throw new Error('Select video host');
		}
		if (!id) {
			throw new Error('Provide video link or ID');
		}
		const res: AxiosResponse<VideoData> = await axios.get(
			`/api/${host}?id=${id}`,
			{ timeout: 10000 }
		);
		return res.data;
	} catch (error: unknown) {
		throw formatFetchError(error);
	}
}

export function createIFrameVideoSource(host: Hosts, id: string): string {
	if (!host) throw new Error('Select video host');
	const embed_domain =
		process.env.NEXT_PUBLIC_NODE_ENV === 'production'
			? process.env.NEXT_PUBLIC_EMBED_DOMAIN
			: process.env.NEXT_PUBLIC_EMBED_DOMAIN_DEV;
	if (!embed_domain) {
		throw new Error('Environment setup error');
	}
	switch (host) {
		case 'youtube':
			return `https://www.youtube-nocookie.com/embed/${id}`;
		case 'youtube-playlist':
			return `https://www.youtube.com/embed/videoseries?list=${id}`;
		case 'twitch':
			return `https://player.twitch.tv/?channel=${id}&parent=${embed_domain}`;
		case 'twitch-vod':
			return `https://player.twitch.tv/?video=${id}&parent=${embed_domain}`;
		case 'dailymotion':
			return `https://www.dailymotion.com/embed/video/${id}`;
		case 'dailymotion-playlist':
			return `https://www.dailymotion.com/embed/playlist/${id}`;
		case 'vimeo':
			return `https://player.vimeo.com/video/${id}`;
		default:
			throw new Error('Unsupported host or incorrect ID');
	}
}

export function createIFrameChatSource(host: Hosts, id: string): string {
	if (!host) throw new Error('Select video host');
	const embed_domain =
		process.env.NEXT_PUBLIC_NODE_ENV === 'production'
			? process.env.NEXT_PUBLIC_EMBED_DOMAIN
			: process.env.NEXT_PUBLIC_EMBED_DOMAIN_DEV;
	if (!embed_domain) {
		throw new Error('Environment setup error');
	}
	switch (host) {
		case 'youtube':
			return `https://www.youtube.com/live_chat?v=${id}&embed_domain=${embed_domain}`;
		case 'twitch':
			return `https://www.twitch.tv/embed/${id}/chat?parent=${embed_domain}`;
		case 'twitch-vod':
			return `https://www.twitch.tv/embed/${id}/chat?parent=${embed_domain}`;
		default:
			throw new Error('Unsupported host or incorrect ID');
	}
}

export function formatFetchError(error: unknown): NextResponse<{
	error: string;
}> {
	if (error instanceof AxiosError) {
		return NextResponse.json(
			{ error: error.response?.data.error || 'Unknown server error' },
			{
				status: error.status || 500,
				statusText: error.response?.data.error || 'Unknown server error',
			}
		);
	} else {
		return NextResponse.json(
			{ error: 'Unknown server error' },
			{ status: 500, statusText: 'Unknown server error' }
		);
	}
}

export function createURLParams(data: VideoData[]): string {
	return (
		'?list=' +
		encodeURIComponent(data.map((v) => v.host + '+' + v.id).join('--'))
	);
}

export async function getDataFromParams(string: string): Promise<VideoData[]> {
	const array = decodeURIComponent(string).split('--');
	const results = await Promise.allSettled(
		array.map(
			(param) =>
				new Promise(async (resolve, reject) => {
					try {
						const str = param.split('+');
						const data = await getVideoData(str[0] as Hosts, str[1]);
						resolve(data);
					} catch (error: any) {
						reject(formatFetchError(error));
					}
				})
		)
	);
	const data = results.map((res) =>
		res.status === 'fulfilled' ? (res.value as VideoData) : null
	);
	return data.filter((r) => r !== null);
}
