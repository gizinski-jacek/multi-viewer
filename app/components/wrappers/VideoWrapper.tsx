import styles from './VideoWrapper.module.scss';
import { VideoData } from '@/app/libs/types';
import { IFrameWrapper } from './IFrameWrapper';
import { createIFrameVideoSource } from '@/app/libs/utils';

interface Props {
	video: VideoData;
	removeVideo: (video: VideoData) => void;
}

export default function VideoWrapper({ video, removeVideo }: Props) {
	return (
		<div className={styles.video}>
			<div
				className={`${styles['remove-video']} btn btn-danger`}
				onClick={() => removeVideo(video)}
			>
				<svg
					width='24px'
					viewBox='0 0 24 24'
					fill='none'
					xmlns='http://www.w3.org/2000/svg'
				>
					<g strokeLinecap='round' strokeLinejoin='round'></g>
					<g>
						<g>
							<path
								d='M18 18L12 12M12 12L6 6M12 12L18 6M12 12L6 18'
								stroke='#000000'
								strokeWidth='2'
								strokeLinecap='round'
								strokeLinejoin='round'
							></path>
						</g>
					</g>
				</svg>
			</div>
			<IFrameWrapper
				src={createIFrameVideoSource(video.host, video.iFrameSrcId)}
				title={`${video.host} video player`}
			/>
		</div>
	);
}
