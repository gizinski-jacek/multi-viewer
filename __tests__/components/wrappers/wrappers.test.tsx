/**
 * @jest-environment jsdom
 */

import { enableFetchMocks } from 'jest-fetch-mock';
enableFetchMocks();
import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import HLSWrapper from '@/components/wrappers/HLSWrapper';
import IFrameWrapper from '@/components/wrappers/IFrameWrapper';
import VideoWrapper from '@/components/wrappers/VideoWrapper';
import { VideoData } from '@/libs/types';
import stylesVideoWrapper from './VideoWrapper.module.scss';
import { createIFrameVideoSource } from '@/libs/utils';

interface Props {
	video: VideoData;
	removeVideo: (video: VideoData) => void;
}

jest.mock('@/components/wrappers/VideoWrapper', () => {
	const VideoWrapper = ({ video, removeVideo }: Props) => (
		<div className={stylesVideoWrapper.video} data-testid='VideoWrapper'>
			<button
				className={`${stylesVideoWrapper['remove-video']} btn btn-danger`}
				type='button'
				onClick={() => removeVideo(video)}
				data-testid='remove-video'
			></button>
			{video.host === 'm3u8' ? (
				<HLSWrapper url={video.id} />
			) : (
				<IFrameWrapper
					src={createIFrameVideoSource(video.host, video.id)}
					title={video.host && `${video.host} video player`}
				/>
			)}
		</div>
	);
	return VideoWrapper;
});

const videoData: VideoData = {
	host: 'youtube',
	id: '9ai3kgu5',
	title: 'Test title',
	channelId: 'y45y5vs1',
	channelName: 'Test name',
	livestreamChat: false,
	thumbnailUrl: null,
};

describe('wrappers', () => {
	describe('HSLWrapper component', () => {
		afterEach(cleanup);
		it('renders video element with default title', () => {
			render(<HLSWrapper url='test' />);
			const video = screen.getByTitle('Stream wrapper');
			expect(video).toHaveProperty('title', 'Stream wrapper');
			expect(video).toBeInTheDocument();
			expect(video).toMatchSnapshot();
		});

		it('renders video element with src property pointing to host url address', () => {
			render(<HLSWrapper url='' />);
			const video = screen.getByTitle('Stream wrapper');
			expect(video).toHaveProperty('src', window.location.origin + '/');
		});

		it('renders video element with src property pointing to provided url', () => {
			const url =
				'https://www.google.com/hls/MjAyNC0xMC0zMFQxMzozMjowMS42NTdac291cHRlbGU/index.m3u8';
			render(<HLSWrapper url={url} />);
			const video = screen.getByTitle('Stream wrapper');
			expect(video).toHaveProperty('src', url);
		});
	});

	describe('IFrameWrapper component', () => {
		afterEach(cleanup);
		it('renders iframe element with default title', () => {
			render(<IFrameWrapper src='test' />);
			const iframe = screen.getByTitle('IFrame wrapper');
			expect(iframe).toHaveProperty('title', 'IFrame wrapper');
			expect(iframe).toBeInTheDocument();
			expect(iframe).toMatchSnapshot();
		});

		it('renders iframe element with custom title and src property pointing to host url address', () => {
			render(<IFrameWrapper src='' title={'Testing unit'} />);
			const iframe = screen.getByTitle('Testing unit');
			expect(iframe).toHaveProperty('title', 'Testing unit');
			expect(iframe).toHaveProperty('src', window.location.origin + '/');
		});

		it('renders iframe element with default title and src property pointing to host url address', () => {
			const src =
				'https://www.google.com/hls/MjAyNC0xMC0zMFQxMzozMjowMS42NTdac291cHRlbGU/index.m3u8';
			render(<IFrameWrapper src={src} />);
			const iframe = screen.getByTitle('IFrame wrapper');
			expect(iframe).toHaveProperty('title', 'IFrame wrapper');
			expect(iframe).toHaveProperty('src', src);
		});
	});

	describe('VideoWrapper component', () => {
		afterEach(cleanup);
		const handleRemove = jest.fn((data) => data);
		const user = userEvent.setup();

		it('renders mocked VideoWrapper component with "video" class with iframe element as last child using mock data', () => {
			render(<VideoWrapper video={videoData} removeVideo={handleRemove} />);
			const video = screen.getByTestId('VideoWrapper');
			expect(video).toBeInTheDocument();
			expect(video).toMatchSnapshot();
			expect(video).toHaveClass('video');
			expect(video.lastChild).toHaveProperty(
				'src',
				createIFrameVideoSource(videoData.host, videoData.id)
			);
		});

		it('renders mocked VideoWrapper component with video element using m3u8 as host service using mock data', () => {
			render(
				<VideoWrapper
					video={{ ...videoData, host: 'm3u8' }}
					removeVideo={handleRemove}
				/>
			);
			const video = screen.getByTestId('VideoWrapper');
			expect(video.lastChild).toHaveProperty(
				'src',
				'http://localhost/' + videoData.id
			);
		});

		it('remove video button renders as first child of VideoWrapper, has click property, and properly registers clicks', () => {
			render(<VideoWrapper video={videoData} removeVideo={handleRemove} />);
			const video = screen.getByTestId('VideoWrapper');
			expect(video.firstChild).toHaveClass('remove-video');
			expect(video.lastChild).toHaveProperty(
				'src',
				createIFrameVideoSource(videoData.host, videoData.id)
			);
			const removeVideo = screen.getByTestId('remove-video');
			expect(removeVideo).toHaveProperty('click');
			expect(handleRemove).toHaveBeenCalledTimes(0);
			user.click(handleRemove(videoData));
			expect(handleRemove).toHaveBeenCalledTimes(1);
		});

		it('remove video function returns mocked data', () => {
			expect(handleRemove).not.toHaveBeenCalled();
			expect(handleRemove(videoData)).toMatchObject(videoData);
			expect(handleRemove).toHaveBeenLastCalledWith(videoData);
		});
	});
});
