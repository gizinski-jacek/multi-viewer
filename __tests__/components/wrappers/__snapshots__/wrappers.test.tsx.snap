// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrappers HSLWrapper component renders video element with default title 1`] = `
<video
  controls=""
  height="100%"
  src="test"
  title="Stream wrapper"
  width="100%"
/>
`;

exports[`wrappers IFrameWrapper component renders iframe element with default title 1`] = `
<iframe
  allow="accelerometer; encrypted-media; gyroscope; picture-in-picture; web-share"
  allowfullscreen=""
  height="100%"
  referrerpolicy="strict-origin-when-cross-origin"
  sandbox="allow-modals allow-scripts allow-same-origin allow-popups allow-popups-to-escape-sandbox allow-storage-access-by-user-activation"
  src="test"
  title="IFrame wrapper"
  width="100%"
/>
`;

exports[`wrappers VideoWrapper component renders mocked VideoWrapper component with "video" class with iframe element as last child using mock data 1`] = `
<div
  class="video"
  data-testid="VideoWrapper"
>
  <button
    class="remove-video btn btn-danger"
    data-testid="remove-video"
    type="button"
  />
  <iframe
    allow="accelerometer; encrypted-media; gyroscope; picture-in-picture; web-share"
    allowfullscreen=""
    height="100%"
    referrerpolicy="strict-origin-when-cross-origin"
    sandbox="allow-modals allow-scripts allow-same-origin allow-popups allow-popups-to-escape-sandbox allow-storage-access-by-user-activation"
    src="https://www.youtube-nocookie.com/embed/9ai3kgu5"
    title="youtube video player"
    width="100%"
  />
</div>
`;
